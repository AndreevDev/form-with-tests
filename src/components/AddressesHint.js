import { useDispatch, useSelector } from "react-redux";
import styled from "styled-components";
import { fetchLocations } from "../actions";

const StyledUl = styled.ul`
    list-style-type: none;
`;

const StyledLi = styled.li`
    cursor: pointer;
    padding: 5px;
`;

const AddressesHint = (props) => {

    const { addressFieldName, setValue } = props;

    const dispatch = useDispatch();

    const { locations } = useSelector(state => state.locations);

    const renderLocations = (locations) => {
        if (!locations || locations.length < 1) {
            return null;
        }
        return (
            locations.map((location, index) => {
                const address = location.address ? location.address.label : location.title;
                return (
                    <StyledLi 
                        key={index}
                        onClick={() => {
                            setValue(addressFieldName, address);
                            dispatch(fetchLocations(""));
                        }}
                    >
                        {address}
                    </StyledLi>
                );
            })
        );
    }

    return (
        <StyledUl>
            {renderLocations(locations)}
        </StyledUl>
    );
}

export default AddressesHint;