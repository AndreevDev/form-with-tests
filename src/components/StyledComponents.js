import styled from "styled-components";

export const Container = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    font-family: Arial, Helvetica, sans-serif !important;
`;

export const StyledForm = styled.form`
    box-shadow: 0px 0px 5px 1px #3c8a123b;
    padding: 20px 30px 10px 30px;
    margin-bottom: 20px;
    width: 25%;
    min-width: 400px;
`;

export const FormHeader = styled.h2`
    margin: 0 0 25px 0;
    text-align: center;
`;

export const ControlWrapper = styled.div`
    position: relative;
    display: flex;
    align-items: center;
    justify-content: ${props => props.justifyContent ? props.justifyContent : 'space-between'};
    margin-bottom: 30px;
    font-size: 1em;
    &:last-child {
        margin-bottom: 15px;
    }

    label {
        padding-right: 10px;
    }

    input, select {
        box-sizing: border-box;
        width: 180px;
        padding: 5px;
        border: ${props => props.errors?.type === 'required' ? "1px solid #f40000" : "1px solid #000000"};
        font-size: 1em;
    }

    input[type="checkbox"] {
        width: auto;
        & + label {
            padding-left: 10px;
            padding-right: 0;
        }
    }

    input[type="submit"] {
        color: green;
        background: transparent;
        border: 1px solid green;
        border-radius: 3px;
        padding: 7px 10px;
        cursor: pointer;
    }
`;

export const StyledError = styled.p`
    color: #f40000;
    margin: 0;
    position: absolute;
    right: 0;
    bottom: -20px;
`;

export const StyledList = styled.ul`
    list-style-type: none;
    
    li {
        cursor: pointer;
        padding: 5px;
    }
`;

export const CenteredP = styled.p`
    text-align: center;
`;

export const StyledTable = styled.table`
    width: 50%;
    min-width: 400px;
    border: 1px solid #000000;
    border-spacing: 0 10px;
    font-size: 17px;
    padding: 0 5px;

    thead tr th {
        border-bottom: 1px solid #000000;
    }

    td {
        padding: 0 10px;
    }

    tbody tr {
        &:nth-child(even) {
            background: #dddddd;
        }
        &:nth-child(odd) {
            background: #ffffff;
        }
    }
`;