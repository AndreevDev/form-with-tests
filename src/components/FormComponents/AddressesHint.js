import { useDispatch, useSelector } from "react-redux";
import { fetchLocations } from "../../actions";
import { StyledList, CenteredP } from "../StyledComponents";

const AddressesHint = (props) => {

    const { addressFieldName, setValue } = props;

    const dispatch = useDispatch();

    const { locations, status } = useSelector(state => state.locations);

    const renderLocations = (locations) => {
        if (!locations || locations.length < 1) {
            return null;
        }

        const items = locations.map((location, index) => {
            const address = location.address ? location.address.label : location.title;
            return (
                <li 
                    key={index}
                    onClick={() => {
                        setValue(addressFieldName, address);
                        dispatch(fetchLocations(""));
                    }}
                >
                    {address}
                </li>
            );
        });

        return (
            <StyledList>
                {items}
            </StyledList>
        );
    }

    const error = status === "error" ? <CenteredP>Error loading locations.</CenteredP> : null;
    const loading = status === "loading" ? <CenteredP>Loading...</CenteredP> : null;
    const content = !(error || loading) ? renderLocations(locations) : null;

    return (
        <>
            {error}
            {loading}
            {content}
        </>
    );
}

export default AddressesHint;