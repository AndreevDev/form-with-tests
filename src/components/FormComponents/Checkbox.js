import React from "react";
import { ControlWrapper, StyledError } from "../StyledComponents";

const Checkbox = React.forwardRef((props, ref) => {
    
    const {errors, ...inputProps} = props;

    return (
        <ControlWrapper errors={errors} justifyContent="start">
            <input type="checkbox" id={inputProps.name} {...inputProps} ref={ref}/>
            <label htmlFor={inputProps.name}>{inputProps.placeholder}</label>
            {errors?.type === 'required' && <StyledError>This field is required</StyledError>}
        </ControlWrapper>
    );
});

export default Checkbox;