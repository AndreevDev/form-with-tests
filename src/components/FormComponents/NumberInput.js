import React from "react";
import { ControlWrapper, StyledError } from "../StyledComponents";

const NumberInput = React.forwardRef((props, ref) => {
    const {errors, ...inputProps} = props;

    return (
        <ControlWrapper errors={errors}>
            <label htmlFor={inputProps.name}>{inputProps.placeholder}</label>
            <input type="number" id={inputProps.name} {...inputProps} ref={ref} 
                onKeyPress={(e) => {
                    if (e.key === "e" || e.key === "+" || e.key === "-") {
                        e.preventDefault();
                    }
                }}/>
            {errors?.type === 'required' && <StyledError>This field is required</StyledError>}
        </ControlWrapper>
    );
});

export default NumberInput;