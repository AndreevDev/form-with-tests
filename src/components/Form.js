import { useEffect } from "react";
import { useForm } from "react-hook-form";
import { useDispatch } from "react-redux";
import { StyledForm, FormHeader } from "./StyledComponents";
import { addCustomer } from "../actions";
import { fetchLocations, locationsCleared } from "../actions";
import { TextInput, NumberInput, Select, AddressesHint, Checkbox, SubmitButton } from "./FormComponents";

const Form = () => {

    const { register, reset, setValue, formState: { errors }, handleSubmit, formState } = useForm();

    const dispatch = useDispatch();

    useEffect(() => {
        if (formState.isSubmitSuccessful) {
            reset({
                firstName: "",
                lastName: "",
                telNumber: "",
                email: "",
                gender: "",
                address: "",
                confirmation: false
            });
          }
    }, [formState, reset]);

    const onSubmit = data => {
        dispatch(addCustomer(data));
        dispatch(locationsCleared());
    };

    const options = [
        {value: "male", label: "Male"},
        {value: "female", label: "Female"}
    ];
    
    return (
        <StyledForm onSubmit={handleSubmit(onSubmit)} noValidate>
            <FormHeader>Registration form</FormHeader>
            <TextInput {...register("firstName", { required: true})}
                placeholder="First Name" 
                errors={errors.firstName} 
            />
            <TextInput {...register("lastName", { required: true })}
                placeholder="Last Name" 
                errors={errors.lastName} 
            />
            <NumberInput {...register("telNumber", { required: true })}
                placeholder="Phone Number" 
                errors={errors.telNumber}
            />
            <TextInput {...register("email", { 
                required: true,
                pattern: {
                    value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
                    message: "Invalid email address"
                } 
            })} 
                type="email" 
                placeholder="Email" 
                errors={errors.email}
            />
            <Select {...register("gender", { required: true })} 
                placeholder="Gender"
                default="Select gender"
                options={options}
                errors={errors.gender}
            />
            <TextInput {...register("address", {
                required: true, 
                onChange: (e) => {
                    if (e.target.value.length > 0) {
                        dispatch(fetchLocations());
                    } else {
                        dispatch(locationsCleared());
                    }
                }
            })}
                placeholder="Address"
                errors={errors.address}
            />
            <AddressesHint addressFieldName="address" setValue={setValue}/>
            <Checkbox {...register("confirmation", { required: true })}
                placeholder="Confirmation"
                errors={errors.confirmation}
            />
            <SubmitButton/>
        </StyledForm>
    );
}

export default Form;