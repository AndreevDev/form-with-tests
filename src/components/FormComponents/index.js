import TextInput from "./TextInput";
import NumberInput from "./NumberInput";
import Select from "./Select";
import Checkbox from "./Checkbox";
import AddressesHint from "./AddressesHint";
import SubmitButton from "./SubmitButton";

export { 
    TextInput,
    NumberInput,
    Select,
    Checkbox,
    AddressesHint,
    SubmitButton
 };