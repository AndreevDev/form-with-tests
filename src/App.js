import Form from "./components/Form";
import Table from "./components/Table";
import { Container } from "./components/StyledComponents";

const App = () => {
    return (
        <Container>
            <Form/>
            <Table/>
        </Container>
    );
}

export default App;
