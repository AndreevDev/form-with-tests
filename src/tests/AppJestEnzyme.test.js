import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import App from "../App";
import { TextInput, NumberInput, Select, Checkbox } from "../components/FormComponents";
import { StyledError } from "../components/StyledComponents";

configure({ adapter: new Adapter() });

describe("App", () => {
    it("App component should render correctly", () => {
        const initialState = {
            customers: [],
            locations: [],
            status: 'idle'
        };
        const mockStore = configureStore();
        const store = mockStore(initialState);
        const tree = renderer.create(<Provider store={store}><App /></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("TextInput", () => {
    it('Should render an TextInput component', () => {
        const wrapper = shallow(<TextInput name="firstName" placeholder="First Name"/>);
        expect(wrapper.find('input[name="firstName"]').exists()).toBe(true);
        expect(wrapper.find('label').text()).toBe("First Name");
    });

    it('Should render an error Required in TextInput component', () => {
        const wrapper = shallow(<TextInput name="lastName" errors={{type: 'required'}}/>);
        expect(wrapper.find(StyledError).text()).toBe('This field is required');
    });

    it('Should render an error Pattern in TextInput component', () => {
        const wrapper = shallow(<TextInput name="email" errors={{type: 'pattern', message: 'Invalid email address'}}/>);
        expect(wrapper.find(StyledError).text()).toBe('Invalid email address');
    });

    it('Should have been changed', () => {
        const onChange = jest.fn();
        const changeEvent = {};
        const wrapper = shallow(<TextInput name="address" onChange={onChange}/>);
        expect(onChange).not.toBeCalled();
        wrapper.find('input').simulate('change', changeEvent);
        expect(onChange).toBeCalledWith(changeEvent);
    });
});

describe("NumberInput", () => {
    it('Should render an TextInput component', () => {
        const wrapper = shallow(<NumberInput name="telNumber" placeholder="Phone Number"/>);
        expect(wrapper.find('input[name="telNumber"]').exists()).toBe(true);
        expect(wrapper.find('label').text()).toBe("Phone Number");
    });

    it('Should render an error Required in TextInput component', () => {
        const wrapper = shallow(<NumberInput name="telNumber" errors={{type: 'required'}}/>);
        expect(wrapper.find(StyledError).text()).toBe('This field is required');
    });

    it('Should have been changed', () => {
        const onChange = jest.fn();
        const changeEvent = {};
        const wrapper = shallow(<NumberInput name="telNumber" onChange={onChange}/>);
        expect(onChange).not.toBeCalled();
        wrapper.find('input').simulate('change', changeEvent);
        expect(onChange).toBeCalledWith(changeEvent);
    });
});

describe("Select", () => {
    const options = [
        {value: "male", label: "Male"},
        {value: "female", label: "Female"}
    ];

    it('Should render an Select component', () => {
        const wrapper = shallow(<Select name="gender" placeholder="Gender" options={options}/>);
        expect(wrapper.find('select[name="gender"]').exists()).toBe(true);
        expect(wrapper.find('label').text()).toBe("Gender");
    });

    it('Should render an error Required in Select component', () => {
        const wrapper = shallow(<Select name="gender" errors={{type: 'required'}} options={options}/>);
        expect(wrapper.find(StyledError).text()).toBe('This field is required');
    });

    it('Should render options correctly', () => {
        const wrapper = shallow(<Select name="gender" errors={{type: 'required'}} options={options}/>);
        expect(wrapper.find('option').at(0).prop('value')).toBe('');
        expect(wrapper.find('option').at(1).prop('value')).toBe('male');
        expect(wrapper.find('option').at(2).prop('value')).toBe('female');
    });
});

describe("Checkbox", () => {
    it('Should render an Checkbox component', () => {
        const wrapper = shallow(<Checkbox name="confirmation" placeholder="Confirmation"/>);
        expect(wrapper.find('input[name="confirmation"]').exists()).toBe(true);
        expect(wrapper.find('label').text()).toBe("Confirmation");
    });

    it('Should render an error Required in Checkbox component', () => {
        const wrapper = shallow(<Checkbox name="confirmation" errors={{type: 'required'}}/>);
        expect(wrapper.find(StyledError).text()).toBe('This field is required');
    });
});