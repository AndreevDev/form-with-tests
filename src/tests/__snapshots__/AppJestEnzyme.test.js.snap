// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App component should render correctly 1`] = `
<div
  className="sc-bdvvtL fyjYjc"
>
  <form
    className="sc-gsDKAQ GdrzK"
    noValidate={true}
    onSubmit={[Function]}
  >
    <h2
      className="sc-dkPtRN jwkkHd"
    >
      Registration form
    </h2>
    <div
      className="sc-hKwDye gXPudc"
    >
      <label
        htmlFor="firstName"
      >
        First Name
      </label>
      <input
        id="firstName"
        name="firstName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="First Name"
        type="text"
      />
    </div>
    <div
      className="sc-hKwDye gXPudc"
    >
      <label
        htmlFor="lastName"
      >
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Last Name"
        type="text"
      />
    </div>
    <div
      className="sc-hKwDye gXPudc"
    >
      <label
        htmlFor="telNumber"
      >
        Phone Number
      </label>
      <input
        id="telNumber"
        name="telNumber"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Phone Number"
        type="number"
      />
    </div>
    <div
      className="sc-hKwDye gXPudc"
    >
      <label
        htmlFor="email"
      >
        Email
      </label>
      <input
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email"
        type="email"
      />
    </div>
    <div
      className="sc-hKwDye gXPudc"
    >
      <label
        htmlFor="gender"
      >
        Gender
      </label>
      <select
        default="Select gender"
        defaultValue=""
        id="gender"
        name="gender"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Gender"
      >
        <option
          disabled={true}
          hidden={true}
          value=""
        >
          Select gender
          ...
        </option>
        <option
          value="male"
        >
          Male
        </option>
        <option
          value="female"
        >
          Female
        </option>
      </select>
    </div>
    <div
      className="sc-hKwDye gXPudc"
    >
      <label
        htmlFor="address"
      >
        Address
      </label>
      <input
        id="address"
        name="address"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Address"
        type="text"
      />
    </div>
    <div
      className="sc-hKwDye jkRzkf"
    >
      <input
        id="confirmation"
        name="confirmation"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Confirmation"
        type="checkbox"
      />
      <label
        htmlFor="confirmation"
      >
        Confirmation
      </label>
    </div>
    <div
      className="sc-hKwDye leWEgE"
    >
      <input
        type="submit"
      />
    </div>
  </form>
  <table
    className="sc-iCfMLu cwlaKr"
  >
    <thead>
      <tr>
        <th>
          First Name
        </th>
        <th>
          Last Name
        </th>
        <th>
          Phone
        </th>
        <th>
          Email
        </th>
        <th>
          Gender
        </th>
        <th>
          Address
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          No customers yet
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
